src/sdk/pynni/nni/compression/torch/builtin_pruners.py

```diff
@@ -104,3 +104,91 @@ def update_epoch(self, epoch):
+    https://arxiv.org/pdf/1811.00250.pdf
+    """
+
+    def __init__(self, config_list):
+        """
+        config_list: supported keys:
+            - pruning_rate: percentage of convolutional filters to be pruned.
+        """
+        super().__init__(config_list)
+        self.mask_list = {}
+
+    def calc_mask(self, weight, config, op, op_type, op_name, **kwargs):
+        """supports Conv1d, Conv2d
+        filter dimensions for Conv1d:
+        IN: number of input channel
+        OUT: number of output channel
+        LEN: filter length
+
+        filter dimensions for Conv2d:
+        IN: number of input channel
+        OUT: number of output channel
+        H: filter height
+        W: filter width
+        """
+
+        assert 0 <= config.get('pruning_rate') < 1
+        assert op_type in ['Conv1d', 'Conv2d']
+        assert op_type in config['op_types']
+
+        if op_name in self.epoch_pruned_layers:
+            assert op_name in self.mask_list
+            return self.mask_list.get(op_name)
+
+        masks = torch.ones(weight.size())
+
+        try:
+            num_kernels = weight.size(0) * weight.size(1)
+            num_prune = int(num_kernels * config.get('pruning_rate'))
+            if num_kernels < 2 or num_prune < 1:
+                return masks
+            min_gm_idx = self._get_min_gm_kernel_idx(weight, num_prune)
+            for idx in min_gm_idx:
+                masks[idx] = 0.
+        finally:
+            self.mask_list.update({op_name: masks})
+            self.epoch_pruned_layers.add(op_name)
+
+        return masks
+
+    def _get_min_gm_kernel_idx(self, weight, n):
+        assert len(weight.size()) in [3, 4]
+
+        dist_list = []
+        for in_i in range(weight.size(0)):
+            for out_i in range(weight.size(1)):
+                dist_sum = self._get_distance_sum(weight, in_i, out_i)
+                dist_list.append((dist_sum, (in_i, out_i)))
+        min_gm_kernels = sorted(dist_list, key=lambda x: x[0])[:n]
+        return [x[1] for x in min_gm_kernels]
+
+    def _get_distance_sum(self, weight, in_idx, out_idx):
+        """ Optimized verision of following naive implementation:
+        def _get_distance_sum(self, weight, in_idx, out_idx):
+            w = weight.view(-1, weight.size(-2), weight.size(-1))
+            dist_sum = 0.
+            for k in w:
+                dist_sum += torch.dist(k, weight[in_idx, out_idx], p=2)
+            return dist_sum
+        """
+        if len(weight.size()) == 4: # Conv2d
+            w = weight.view(-1, weight.size(-2), weight.size(-1))
+            anchor_w = weight[in_idx, out_idx].unsqueeze(0).expand(w.size(0), w.size(1), w.size(2))
+        elif len(weight.size()) == 3: # Conv1d
+            w = weight.view(-1, weight.size(-1))
+            anchor_w = weight[in_idx, out_idx].unsqueeze(0).expand(w.size(0), w.size(1))
+        else:
+            raise RuntimeError('unsupported layer type')
```

Comment:
print the weight.size() here? For get more debug information.

Reply:
Thanks,  debug info added.